### ISSUE TITLE
font-label counts from 1, when the fonts count from 0.

### ISSUE THREAD

**Issue by 132ikl** [jaagr/polybar#1305]
In my config, let's say I have this:
```
font-0 = "mycoolfont:12"
font-1 = "myawesomefont:14"
```
If you try to apply this font to a label, let's say in
```
[custom/mycoolscript]
interval = 10
exec = /path/to/script
label-font = 1
```
You would expect this to use font-1, my awesome font. But instead, this uses font-0, my cool font.

**Comment by szymonmaszke** [comment]
It is actually described in polybar's documentation IIRC, this change would break backwards compatibility and configs of many people.

**Comment by 132ikl** [comment]
It is [here.](https://github.com/jaagr/polybar/wiki/Fonts)  Still, it can be very confusing for novice users who are trying to implement a basic script without delving too far into custom configuration. Maybe a flag could be introduced that is generated on the default config to fix the issue, while also not breaking backwards compatibility?

**Comment by NBonaparte** [comment]
I agree it is somewhat confusing. We can probably add something like that for the next release.

**Comment by patrick96** [comment]
I think, the original intention behind this, is to ensure compatibility with lemonbar, because back in the day, polybar would only generate the input for lemonbar and not create the bar itself. lemonbar uses a 1-based index for the font tags. I agree this is *very* confusing, even for me sometimes, so a config option would be something we should probably do.

**Comment by carymrobbins** [comment]
Maybe we could add a new config entry to enable zero-based access?

`zero-indexed-fonts = true`

**Comment by x70b1** [comment]
I would like to see it in `Polybar 2.0.0`.

**Comment by patrick96** [comment]
Just writing down some thoughts right now, for when we try to implement this.

The font can be changed by both the `-font` property and the `%{T}` tag. We'll probably want to be consistent and if zero-indexed fonts are enabled, both `-font` and `%{T}` should use them. This means all `%{T}` tags generated by polybar will need to use zero-indexed fonts. And that either the parser needs to change the index before sending the `change_font` event to the renderer, the renderer needs to adjust the font number or even the `context& operator<<(const textblock& t)` function in `context.hpp` which does the font prioritization needs to change its functionality. I don't like the last option because we would need to bring configuration options into unrelated code. In the renderer seems best for me.

We could also use this opportunity to change how `context& operator<<(const textblock& t)`  does font prioritization, right now it just switches the first font with the desired font in the font list. Under some circumstances this can lead to some weird behaviour, it probably should be that the preferred font is just moved to the front and the others pushed back.

**Comment by 132ikl** [comment]
Any update for this issue?

**Comment by Lomadriel** [comment]
> We could also use this opportunity to change how `context& operator<<(const textblock& t)` does font prioritization, right now it just switches the first font with the desired font in the font list. Under some circumstances this can lead to some weird behaviour, it probably should be that the preferred font is just moved to the front and the others pushed back.

I am working on this feature but I would like to confirm how the font priority works.
```ini

[bar/custom]
font-0 = …
font-1 = …
font-2 = …
font-3 = …

[module/test]
label-font = 3
```
If I am not wrong when treating the "test" module, the priority order used by the function `context& operator<<(const textblock& t)` should be (the following list is 0 indexed):
1. font-2
2. font-0
3. font-1
4. font-3

Or am I missing something?

**Comment by patrick96** [comment]
@Lomadriel no, this approach is correct, we move the desired font in front.

**Comment by Lomadriel** [comment]
@patrick96 
But what is the new priority order you wanted?

> We could also use this opportunity to change how `context& operator<<(const textblock& t)` does font prioritization, right now it just switches the first font with the desired font in the font list. Under some circumstances this can lead to some weird behaviour, it probably should be that the preferred font is just moved to the front and the others pushed back.

**Comment by patrick96** [comment]
Sorry. In your example the current implementation generates the following font order:

1. font-2
1. font-1
1. font-0
1. font-3

(it swaps `font-2` and `font-0`). What we want is the one you described, where the target font is moved to the front and the others pushed back.

**Comment by Lomadriel** [comment]
Okay, I'll start implementing this ^^
